kotlin/src/main/kotlin/org/axonframework/extensions/kotlin/serialization/KotlinSerializer.kt

```diff
@@ -0,0 +1,205 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.axonframework.extensions.kotlin.serialization
+
+import kotlinx.serialization.KSerializer
+import kotlinx.serialization.SerializationException
+import kotlinx.serialization.json.Json
+import kotlinx.serialization.json.JsonElement
+import kotlinx.serialization.json.JsonNull
+import org.axonframework.common.ObjectUtils
+import org.axonframework.serialization.AnnotationRevisionResolver
+import org.axonframework.serialization.ChainingConverter
+import org.axonframework.serialization.Converter
+import org.axonframework.serialization.RevisionResolver
+import org.axonframework.serialization.SerializedObject
+import org.axonframework.serialization.SerializedType
+import org.axonframework.serialization.Serializer
+import org.axonframework.serialization.SimpleSerializedObject
+import org.axonframework.serialization.SimpleSerializedType
+import org.axonframework.serialization.UnknownSerializedType
+import kotlin.reflect.full.companionObject
+import kotlin.reflect.full.companionObjectInstance
+
+/**
+ * Implementation of Axon Serializer that uses a kotlinx.serialization implementation.
+ * The serialized format is JSON.
+ *
+ * The DSL function kotlinSerializer can be used to easily configure the parameters
+ * for this serializer.
+ *
+ * @see kotlinx.serialization.Serializer
+ * @see org.axonframework.serialization.Serializer
+ * @see kotlinSerializer
+ */
+class KotlinSerializer(
+    private val revisionResolver: RevisionResolver,
+    private val converter: Converter,
+    private val json: Json
+) : Serializer {
+
+    private val serializerCache: MutableMap<Class<*>, KSerializer<*>> = mutableMapOf()
+
+    override fun <T> serialize(value: Any?, expectedRepresentation: Class<T>): SerializedObject<T> {
+        try {
+            val type = ObjectUtils.nullSafeTypeOf(value)
+            return when {
+                expectedRepresentation.isAssignableFrom(String::class.java) ->
+                    SimpleSerializedObject(
+                        (if (value == null) "null" else json.encodeToString(type.serializer(), value)) as T,
+                        expectedRepresentation,
+                        typeForClass(type)
+                    )
+
+                expectedRepresentation.isAssignableFrom(JsonElement::class.java) ->
+                    SimpleSerializedObject(
+                        (if (value == null) JsonNull else json.encodeToJsonElement(type.serializer(), value)) as T,
+                        expectedRepresentation,
+                        typeForClass(type)
+                    )
+
+                else ->
```

Comment:
[]I would've assumed the `Serializer` could default to `byte[]` instead of throwing an exception.
Doing so will have this serializer follow the same paradigm as the `JacksonSerializer` and `XStreamSerializer`.

Several spots in the framework assume that the `expectedRepresentation` is `byte[]`.
So enabling this option up would make this serializer more valuable throughout the framework.

Reply:
Pretty confident this could this will require introducing the `registerConverters` method (as seen on line 128 of the `JacksonSerializer` too).
This method ensures Json specific `ContentTypeConverters` are added to the set-up. 

Furthermore, I assume this will warrant introducing content type converters going from `JsonElement` to `byte[]`, and vice versa.

Reply:
I totally forgot about the (extra) converters. I will add some defaults to convert between byte arrays <-> Strings and JSON Nodes :+1:

Reply:
I made the `JsonElement` type the exception, just like `JsonNode` in the Jackson serializer. All other target types are serialized to String, and then converted with the chaining converter.

By default String, Byte array, Json Element and InputStream already work through service discovery, so that is fine.

If any other converters are needed I can add them.

Reply:
Thanks for validating @hiddewie :+1: 
Adding tests to validate this works (as in, the conversion between those types through this serializer) are a massive plus.